lib/Common/Memory/RecyclerTelemetryInfo.cpp

```diff
@@ -0,0 +1,256 @@
+        this->FreeGCPassStats();
+    }
+
+    void RecyclerTelemetryInfo::FillInSizeData(IdleDecommitPageAllocator* allocator, AllocatorSizes* sizes) const
+    {
+        sizes->committedBytes = allocator->GetCommittedBytes();
+        sizes->reservedBytes = allocator->GetReservedBytes();
+        sizes->usedBytes = allocator->GetUsedBytes();
+        sizes->numberOfSegments = allocator->GetNumberOfSegments();
+    }
+
+    void RecyclerTelemetryInfo::StartPass()
+    {
+        Js::Tick start = Js::Tick::Now();
+        if(!this->captureTelemetryData)
+        { 
+            return;
+        }
+
+        //Output::Print(L"startPass called (current thread = %p, main thread = %p), passCount  = %i\r\n", GetCurrentThreadContextId(), this->mainThreadID, this->passCount);
+        AssertOnMainThread(this, RecyclerTelemetryInfo::StartPass);
+#if DBG
+        // validate state of existing GC pass stats structs
+        uint16 count = 0;
+        if (this->lastPassStats != nullptr)
+        {
+            RecyclerTelemetryGCPassStats* head = this->lastPassStats->next;
+            RecyclerTelemetryGCPassStats* curr = head;
+            do
+            {
+                AssertMsg(curr->isGCPassActive == false, "unexpected value for isGCPassActive");
+                count++;
+                curr = curr->next;
+            } while (curr != head);
+        }
+        AssertMsg(count == this->passCount, "RecyclerTelemetryInfo::StartPass() - mismatch between passCount and count.");
+#endif
+
+        RecyclerTelemetryGCPassStats* p = HeapNewNoThrow(RecyclerTelemetryGCPassStats);
+        if (p == nullptr)
+        {
+            // failed to allocate memory - disable any further telemetry capture for this recycler 
+            // and free any existing GC stats we've accumulated
+            this->captureTelemetryData = false;
+            FreeGCPassStats();
+            this->hostInterface->TransmitTelemetryError(*this, "Memory Allocation Failed");
+        }
+        else
+        {
+            passCount++;
+            memset(p, 0, sizeof(RecyclerTelemetryGCPassStats));
+            if (this->lastPassStats == nullptr)
+            {
+                p->next = p;
+            }
+            else
+            {
+                p->next = lastPassStats->next;
+                this->lastPassStats->next = p;
+            }
+            this->lastPassStats = p;
+
+            this->lastPassStats->isGCPassActive = true;
+            this->lastPassStats->passStartTimeTick = Js::Tick::Now();
+            GetSystemTimePreciseAsFileTime(&this->lastPassStats->passStartTimeFileTime);
+            if (this->hostInterface != nullptr)
+            {
+                LPFILETIME ft = this->hostInterface->GetLastScriptExecutionEndTime();
+                this->lastPassStats->lastScriptExecutionEndTime = *ft;
+            }
+
+            this->lastPassStats->processCommittedBytes_start = RecyclerTelemetryInfo::GetProcessCommittedBytes();
```

Comment:
Do we care about dealing with the case where this call fails and returns 0?

Reply:
I don't think we need to worry about this.  If this fails for some reason, it will show up in telemetry data that we have a process with 0 committed bytes.